File: code/update_final_results.R
update_final_results <- function(method, table_id, rmse = NA, nlpd = NA, runtime = NA,
                                 version = "x",
                                 final_csv = file.path("results", "final results.csv")) {
  if (!file.exists(final_csv)) {
    stop("final results.csv does not exist. Please create the template first.")
  }
  
  df <- read.csv(final_csv, header = FALSE, stringsAsFactors = FALSE, na.strings = c("", "NA"))
  df[is.na(df)] <- ""
  
  find_row <- function(table_label, method_name) {
    table_row <- which(df[, 1] == table_label)
    if (length(table_row) == 0) stop(paste("Could not find", table_label))
    
    next_table_row <- which(df[, 1] %in% c("Table 2", "Table 3", "Table 4") & seq_len(nrow(df)) > table_row[1])
    end_row <- if (length(next_table_row) == 0) nrow(df) else next_table_row[1] - 1
    
    method_rows <- which(df[, 1] == method_name)
    method_rows <- method_rows[method_rows > table_row[1] & method_rows <= end_row]
    if (length(method_rows) == 0) stop(paste("Could not find method", method_name, "under", table_label))
    
    method_rows[1]
  }
  
  r <- find_row(table_id, method)
  
  if (table_id %in% c("Table 2", "Table 3")) {
    df[r, 2] <- ifelse(is.na(rmse),    "", sprintf("%.2f", rmse))
    df[r, 3] <- ifelse(is.na(nlpd),    "", sprintf("%.2f", nlpd))
    df[r, 4] <- ifelse(is.na(runtime), "", sprintf("%.2f", runtime))
  } else if (table_id == "Table 4") {
    if (version == "x")       df[r, 2] <- ifelse(is.na(rmse), "", sprintf("%.2f", rmse))
    else if (version == "xs") df[r, 3] <- ifelse(is.na(rmse), "", sprintf("%.2f", rmse))
    else stop("version must be 'x' or 'xs'")
  } else {
    stop("table_id must be one of: Table 2, Table 3, Table 4")
  }
  
  write.table(df, final_csv, sep = ",", row.names = FALSE, col.names = FALSE, quote = TRUE)
}

File: code/update_final_results.py
import pandas as pd
import numpy as np
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

def update_final_results(method, table_id, rmse=np.nan, nlpd=np.nan, runtime=np.nan,
                         version="x", final_csv=None):
    """
    version : "x"  -> writes to RMSE(x)   column (Table 4 only)
              "xs" -> writes to RMSE(x+s)  column (Table 4 only)
              ignored for Table 2 and Table 3
    """
    if final_csv is None:
        final_csv = REPO_ROOT / "results" / "final results.csv"
    else:
        final_csv = Path(final_csv)

    if not final_csv.exists():
        raise FileNotFoundError(f"{final_csv} does not exist. Please create the template first.")

    df = pd.read_csv(final_csv, header=None, dtype=str).fillna("")

    def find_row(table_label, method_name):
        table_rows = df.index[df[0] == table_label].tolist()
        if not table_rows:
            raise ValueError(f"Could not find {table_label}")
        table_row = table_rows[0]
        next_table_rows = df.index[
            (df[0].isin(["Table 2", "Table 3", "Table 4"])) & (df.index > table_row)
        ].tolist()
        end_row = len(df) - 1 if not next_table_rows else next_table_rows[0] - 1
        method_rows = df.index[df[0] == method_name].tolist()
        method_rows = [r for r in method_rows if table_row < r <= end_row]
        if not method_rows:
            raise ValueError(f"Could not find method '{method_name}' under {table_label}")
        return method_rows[0]

    r = find_row(table_id, method)
    fmt = lambda x: "" if pd.isna(x) else f"{x:.2f}"

    if table_id in ["Table 2", "Table 3"]:
        df.iat[r, 1] = fmt(rmse)
        df.iat[r, 2] = fmt(nlpd)
        df.iat[r, 3] = fmt(runtime)
    elif table_id == "Table 4":
        if version == "x":
            df.iat[r, 1] = fmt(rmse)
        elif version == "xs":
            df.iat[r, 2] = fmt(rmse)
        else:
            raise ValueError("version must be 'x' or 'xs'")
    else:
        raise ValueError("table_id must be one of: Table 2, Table 3, Table 4")

    df.to_csv(final_csv, header=False, index=False)
